OpenRA.Mods.Common/Activities/FindResources.cs

```diff
@@ -50,51 +58,80 @@ public override Activity Tick(Actor self)
 			var searchRadius = harv.LastOrderLocation.HasValue ? harvInfo.SearchFromOrderRadius : harvInfo.SearchFromProcRadius;
-			var searchRadiusSquared = searchRadius * searchRadius;
 
-			// Find harvestable resources nearby:
-			var path = self.World.WorldActor.Trait<IPathFinder>().FindPath(
-				PathSearch.Search(self.World, mobileInfo, self, true)
-					.WithHeuristic(loc =>
+			// Pick the closest harvestable positions inside the radius from the selected point.
+			// This point will be used as a fallback in case no closer resources are found
+			var closestHarvestablePosition = self.World.Map.FindTilesInAnnulus(searchFromLoc, 0, searchRadius)
+				.FirstOrDefault(loc =>
 					{
-						// Avoid this cell:
-						if (avoidCell.HasValue && loc == avoidCell.Value)
-							return EstimateDistance(loc, searchFromLoc) + Constants.CellCost;
-
-						// Don't harvest out of range:
-						var distSquared = (loc - searchFromLoc).LengthSquared;
-						if (distSquared > searchRadiusSquared)
-							return EstimateDistance(loc, searchFromLoc) + Constants.CellCost * 2;
-
-						// Get the resource at this location:
 						var resType = resLayer.GetResource(loc);
 						if (resType == null)
-							return EstimateDistance(loc, searchFromLoc) + Constants.CellCost;
+							return false;
 
 						// Can the harvester collect this kind of resource?
 						if (!harvInfo.Resources.Contains(resType.Info.Name))
-							return EstimateDistance(loc, searchFromLoc) + Constants.CellCost;
+							return false;
 
 						if (territory != null)
 						{
 							// Another harvester has claimed this resource:
 							ResourceClaim claim;
-							if (territory.IsClaimedByAnyoneElse(self, loc, out claim))
-								return EstimateDistance(loc, searchFromLoc) + Constants.CellCost;
+							if (territory.IsClaimedByAnyoneElse(self, loc, out claim)) return false;
 						}
 
-						return 0;
-					})
-					.FromPoint(self.Location));
+						return true;
+					});
 
-			if (path.Count == 0)
+			List<CPos> path;
+
+			// If no harvestable position could be found, then return an empty path
+			if (closestHarvestablePosition != default(CPos))
+			{
+				// Find a path to that harvestable resource
+				path = self.World.WorldActor.Trait<IPathFinder>()
+					.FindPath(PathSearch.Search(self.World, mobileInfo, self, true)
+						.WithHeuristic(loc =>
+						{
+							var resType = resLayer.GetResource(loc);
+							if (resType == null)
+								return DefaultEstimator(loc, closestHarvestablePosition);
+
+							// Can the harvester collect this kind of resource?
+							if (!harvInfo.Resources.Contains(resType.Info.Name))
+								return DefaultEstimator(loc, closestHarvestablePosition);
+
+							if (territory != null)
+							{
+								// Another harvester has claimed this resource:
+								ResourceClaim claim;
+								if (territory.IsClaimedByAnyoneElse(self, loc, out claim))
+									return DefaultEstimator(loc, closestHarvestablePosition);
+							}
+
+							return 0;
```

Comment:
It would be worth extracting the delegate above used to determine the ```closestHarvestablePosition``` into a function and reusing the logic here to prevent the duplication. e.g.
```WithHeuristic(loc => CanHarvest(loc) ? 0 : DefaultEstimator(loc, closestHarvestablePosition));```

(unless this is inlining for perf reasons or whatever)